src/v/kafka/server/handlers/create_topics.cc

```diff
@@ -344,7 +345,8 @@ ss::future<response_ptr> create_topics_handler::handle(
       });
     valid_range_end = quota_exceeded_it;
 
-    auto to_create = to_cluster_type(begin, valid_range_end);
```

Comment:
In this commit you are not changing just the interface, you also changing the response, right? Or there are no "functional changes" because the to_cluster_type method can't fail/throw?

Reply:
I'm not worried about splitting it further. I'm trying to understand the change in behavior. Would be great to note it in the commit message for the next person too. Ideally a test as well that shows what changed. Did we previously abort the whole create topics operation? Do we partially create topics now? Want to have a clear understanding of the changes we are merging mostly.

Reply:
The previous behavior would be that the uncaught exception from a bad `boost::lexical_cast()` would be propagated up to the `request_context` layer, where it would be retried until a timeout- bad UX experience leading to wholly aborted `CreateTopics` requests and an `rpk` output that doesn't accurately describe the issue.
e.g,

`$ rpk topic create tapioca -c retention.bytes=this_should_be_an_integer`

`redpanda` output:

```
WARN  2024-10-09 13:24:15,329 [shard  1:main] kafka - connection_context.cc:872 - Error processing request: boost::wrapexcept<boost::bad_lexical_cast> (bad lexical cast: source type value could not be interpreted as target)
WARN  2024-10-09 13:24:15,626 [shard  1:main] kafka - connection_context.cc:872 - Error processing request: boost::wrapexcept<boost::bad_lexical_cast> (bad lexical cast: source type value could not be interpreted as target)
WARN  2024-10-09 13:24:16,081 [shard  1:main] kafka - connection_context.cc:872 - Error processing request: boost::wrapexcept<boost::bad_lexical_cast> (bad lexical cast: source type value could not be interpreted as target)
WARN  2024-10-09 13:24:16,914 [shard  1:main] kafka - connection_context.cc:872 - Error processing request: boost::wrapexcept<boost::bad_lexical_cast> (bad lexical cast: source type value could not be interpreted as target)
WARN  2024-10-09 13:24:19,145 [shard  1:main] kafka - connection_context.cc:872 - Error processing request: boost::wrapexcept<boost::bad_lexical_cast> (bad lexical cast: source type value could not be interpreted as target)
WARN  2024-10-09 13:24:21,654 [shard  1:main] kafka - connection_context.cc:872 - Error processing request: boost::wrapexcept<boost::bad_lexical_cast> (bad lexical cast: source type value could not be interpreted as target)
WARN  2024-10-09 13:24:24,164 [shard  1:main] kafka - connection_context.cc:872 - Error processing request: boost::wrapexcept<boost::bad_lexical_cast> (bad lexical cast: source type value could not be interpreted as target)
```

`rpk` output:

```
unable to create topics [topic]: broker closed the connection immediately after a request was issued, which happens when SASL is required but not provided: is SASL missing?
```

> Did we previously abort the whole create topics operation? Do we partially create topics now?

Previously, we would abort the whole create topics operation, and after the changes in this PR, that is still the behaviour. However, the UX is improved through the caught exception and error code response.

I can add a test that demonstrates the "all-or-nothing" behaviour for creating multiple topics (some with valid configs, some with invalid).

Reply:
@WillemKauf the way I read the code is that the behavior now changed. You are populating the `response.data.topics` with errors for things which fail the validation but then continue with create_topics() function call with return value (`to_create`).

If what I'm saying is true, shall we double check with kafka behavior? My intuition is that new behavior (the way I describe it) should be fine given that we proceed somewhat similarly when quote is exceeded (validation above; continue with what fits inside the quota)

Reply:
Hm. My tests seemed to reflect all-or-nothing behavior but https://kafka.apache.org/20/javadoc/org/apache/kafka/clients/admin/AdminClient.html#createTopics-java.util.Collection-org.apache.kafka.clients.admin.CreateTopicsOptions- states that `this operation is not transactional so it may succeed for some topics while fail for others`.

Double checking why this is the case.

Reply:
Ah my test had a bug in it. Indeed, we do partially create the valid topics.

Reply:
Made a custom build of rpk which allows custom configs for different topics (we don't support it by default) and just confirming kafka behavior:

```
 → ~/redpanda/vbuild/go/linux/amd64/bin/rpk topic create foo bar baz
TOPIC  STATUS
foo    INVALID_CONFIG: Invalid value foo for configuration unclean.leader.election.enable: Expected value to be either true or false
bar    INVALID_CONFIG: Invalid value bar for configuration unclean.leader.election.enable: Expected value to be either true or false
baz    OK
```

Partial success. 👍